dandi/dandiapi.py

```diff
@@ -273,11 +276,26 @@ def dandi_authenticate(self):
             raise NotImplementedError("TODO client name derivation for keyring")
         app_id = f"dandi-api-{client_name}"
         keyring_backend, api_key = keyring_lookup(app_id, "key")
-        if not api_key:
-            api_key = input(f"Please provide API Key for {client_name}: ")
-            keyring_backend.set_password(app_id, "key", api_key)
-            lgr.debug("Stored key in keyring")
-        self.authenticate(api_key)
+        key_from_keyring = api_key is not None
+        while True:
+            if not api_key:
+                api_key = input(f"Please provide API Key for {client_name}: ")
+                key_from_keyring = False
+            try:
+                self.authenticate(api_key)
+            except requests.HTTPError:
+                if is_interactive() and click.confirm(
+                    "API key is invalid; enter another?"
+                ):
+                    api_key = None
+                    continue
+                else:
+                    raise
+            else:
+                if not key_from_keyring:
```

Comment:
why do you think it is worth to not store a new key if we got it from a keyring but it is no longer valid?  I would just always store it

Reply:
If a keyring key is invalid, `key_from_keyring` will be set to `False` after the user enters a new key, so the new key will be stored in the keyring.

Reply:
oh right -- so we store only if a new key was entered, got it, thanks